flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/SourceReaderBase.java

```diff
@@ -99,67 +106,94 @@ public SourceReaderBase(
 		this.splitFetcherManager = splitFetcherManager;
 		this.recordEmitter = recordEmitter;
 		this.splitStates = new HashMap<>();
-		this.splitIter = null;
 		this.options = new SourceReaderOptions(config);
 		this.config = config;
 		this.context = context;
 		this.noMoreSplitsAssignment = false;
 	}
 
 	@Override
-	public void start() {
-
-	}
+	public void start() {}
 
 	@Override
 	public InputStatus pollNext(ReaderOutput<T> output) throws Exception {
 		splitFetcherManager.checkErrors();
-		// poll from the queue if the last element was successfully handled. Otherwise
-		// just pass the last element again.
-		RecordsWithSplitIds<E> recordsWithSplitId = null;
-		boolean newFetch = splitIter == null || !splitIter.hasNext();
-		if (newFetch) {
-			recordsWithSplitId = elementsQueue.poll();
+
+		// make sure we have a fetch we are working on, or move to the next
```

Comment:
Thanks for the refactoring, the logic is much clearer now.

Reply:
Thanks :-)